Add annual software plans as default product plans

nospame opened the pull request:
## Product Description
Adds "Pay Annually" software plans if they do not already exist, using the naming convention "CommCare {edition} Edition - Pay Annually".

## Technical Summary
[SaaS-15493](https://dimagi.atlassian.net/browse/SAAS-15493)
Review by commit recommended.

## Feature Flag
This change is not feature flagged.

## Safety Assurance

### Safety story
Tested migration directly in local environment and on staging, confirming new plans have been created or modified as expected and not duplicated.
Because `ensure_plans` is always run via migrations, these changes are tested implicitly every time db test setup runs. 

### Automated test coverage
Added config for annual plans and new asserts in `test_ensure_plans`.

### QA Plan
Given complexity I ran into locally in constructing these changes, I would consider QA here, but I am not sure how to do so since it is primarily a data migration. Any thoughts?

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

The migration here relies on code changes to `ensure_plans` included in this PR to run correctly.

### Rollback instructions
The migration should be reversed first. This would remove `DefaultProductPlan` objects added here but _not_ any `SoftwarePlan`s, because on some environments those annual plans would have already existed.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Refactor ensure_plans

- Break down into more model-specific _ensure functions
- namedtuple for plan_key parameters
- get_or_create for SoftwarePlan
- helpers for plan name and SoftwareProduct
- kwargs dict for SoftwarePlan and DefaultProductPlan

nospame committed:
Allow annual DefaultProductPlan instances

Adding is_annual_plan to unique_together constraint means the model can support a monthly and an annual version of what otherwise looks like the same plan

nospame committed:
Update ensure_plans to support annual plans

nospame committed:
Bootstrap annual software plans via data migration

nospame committed:
Update test_ensure_plans with is_annual_plan positional arg

nospame committed:
Add asserts for is_annual_plan and visibility set correctly

nospame committed:
Add annual software plans to testing config

nospame committed:
Update migrations.lock

jingcheng16 commented on corehq/apps/accounting/bootstrap/config/testing.py:
```diff
@@ -7,15 +7,15 @@
 
 BOOTSTRAP_CONFIG_TESTING = {
-    (SoftwarePlanEdition.COMMUNITY, False, False): {
+    (SoftwarePlanEdition.COMMUNITY, False, False, False): {
```
I guess the test won't fail if you didn't update those configs since you have made changes to `ensure_plans` to account for old config file who don't have `is_annual_plan`. Can you validate it ?( by checking out a previous commit and run the test)

nospame replied:
I found it was failing, and I think that once `DefaultProductPlan` has `is_annual_plan` property on it, added to its `unique_together` constraint, and objects created using `is_annual_plan=True`, that key should always be passed in. 

The reason being that in `ensure_plans`, we try to get a `DefaultProductPlan` directly based on whether or not that key exists. If we leave out that key at this point, `get()` can return multiple objects (one with `is_annual_plan=True` and one `is_annual_plan=False`), which throws an unexpected error. 
https://github.com/dimagi/commcare-hq/blob/486bd7811aa2362f2d79852e8ca47dbc65aeeef0/corehq/apps/accounting/bootstrap/utils.py#L217-L221

So, essentially that key is optional in the past, but is required going forward.

jingcheng16 reviewed (commented):
Looks good! Thank for the refactoring. Just a few questions.

nospame commented on corehq/apps/accounting/bootstrap/config/annual_plans_may_2024.py:
```diff
@@ -0,0 +1,30 @@
+
+from corehq.apps.accounting.models import FeatureType, SoftwarePlanEdition
+
+BOOTSTRAP_CONFIG = {
```
Yes, I've confirm running locally that this generates plans named "CommCare [Standard/Pro/Advanced] Edition - Pay Annually" which matches what we have on production.

nospame commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +162,69 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+        'edition': plan_key.edition,
+        'visibility': SoftwarePlanVisibility.PUBLIC
+    }
+    software_plan, created = SoftwarePlan.objects.get_or_create(**plan_opts)
```
That's a good point, and I can't tell which is better here.

Previously, we would instantiate a new object with these parameters, then try to look one up in the db based on the name. If we find it based on its name, which is required to be unique between plans, we're basically throwing away `edition` and `visibility`. If we don't find it based on its name, we create a new plan using all three parameters.

Now, we will try to find an object in the db based on all three parameters, and if it doesn't match we create a new plan using the parameters. If the new object tries to use the same name as an existing plan, we will get an error, because name must be unique.

The only practical difference in behavior I see is if, for some reason we pass in a plan that has the same name but a different `edition` and/or `visibility` than an existing plan. Previously, we'd silently ignore those keys. Now, we'll get an error that name must be unique.

nospame replied:
On second thought, while I like the cleanness of `get_or_create` I think it's unnecessary to construct and pass in all the options here. Since we already know the name must be unique, we can try to `get` the plan first, then if that fails by `DoesNotExist`, simply `create` the new plan. That will retain existing behavior while structuring to be a little neater.

nospame committed:
Try to get software plan by name first

No need to pass in all params when name is sufficient. Reverts to previous behavior where visibility and edition are silently ignored if they mismatch the existing plan.

nospame committed:
Update SoftwareProduct comment

nospame commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +162,69 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+        log_accounting_info("Creating Software Plan: %s" % software_plan.name)
+    if verbose and not created:
+        log_accounting_info(
+            "Plan '%s' already exists. Using existing plan to add version." % software_plan.name
```
I opted to revert to previous behavior instead. https://github.com/dimagi/commcare-hq/pull/34584/commits/f8321bd2f321ccfc41596e1dfed6e7b97d04c5f9

biyeun commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +163,84 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+
+def _software_plan_name(plan_key, product, product_rate):
+    plan_name = (
+        (('%s Trial' % product_rate.name) if plan_key.is_trial else ('%s Edition' % product_rate.name))
```
I know this is copied from very old code, but can we use f-strings to make this more readable?
```suggestion
        (f'{product_rate.name} Trial' if plan_key.is_trial else f'{product_rate.name} Edition')
```

millerdev replied:
Could this need to be translated, or is this always used for internal representations like `log_accounting_info`? If it may be translated then f-strings cannot be used, although `string.format()` is still better than the old `%` syntax in most cases.

nospame replied:
As far as I can see it's only ever used internally here, since we're using it to look up and/or generate model instances. Since it does get shown to a user in limited cases like renewing their plan, it could be that it's translated later on (I don't think so, but it could be), but I would think that would be based on the string saved on the model instance, not this string construction.

biyeun commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +163,84 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+        (('%s Trial' % product_rate.name) if plan_key.is_trial else ('%s Edition' % product_rate.name))
+        if product is None else product.name)
+    if plan_key.is_annual_plan:
+        plan_name = '%s - Pay Annually' % plan_name
```
continuing f string comment above
```suggestion
        plan_name = f'{plan_name} - Pay Annually'
```

biyeun commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +163,84 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+    if plan_key.is_annual_plan:
+        plan_name = '%s - Pay Annually' % plan_name
+    if plan_key.is_report_builder_enabled:
+        plan_name = '%s - Report Builder (5 Reports)' % plan_name
```
```suggestion
        plan_name = f'{plan_name} - Report Builder (5 Reports)'
```

biyeun commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +163,84 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+            plan_opts['is_annual_plan'] = plan_key.is_annual_plan
+        software_plan = SoftwarePlan.objects.create(**plan_opts)
+        if verbose:
+            log_accounting_info("Creating Software Plan: %s" % software_plan.name)
```
```suggestion
            log_accounting_info(f"Creating Software Plan: {software_plan.name}")
```

biyeun commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -123,32 +73,37 @@ def _ensure_product_rate(monthly_fee, edition, verbose, apps):
+        product = SoftwareProduct.objects.get(name=product.name)
+        if verbose:
+            log_accounting_info(
+                "Product '%s' already exists. Using existing product to add rate."
+                % product.name
```
```suggestion
                f"Product '{product.name}' already exists. Using existing product to add rate."
```

biyeun commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +163,84 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+        default_product_plan.save()
+        if verbose:
+            log_accounting_info(
+                "Setting plan as default for edition '%s' with is_trial='%s'."
+                % (default_product_plan.edition, plan_key.is_trial)
```
```suggestion
                f"Setting plan as default for edition '{default_product_plan.edition}' "
                f"with is_trial='{plan_key.is_trial}'."
```

biyeun commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -208,6 +163,84 @@ def _ensure_feature_rates(feature_rates, features, edition, verbose, apps):
+        default_product_plan = DefaultProductPlan.objects.get(**plan_opts)
+        if verbose:
+            log_accounting_info(
+                "Default for edition '%s' with is_trial='%s' already exists."
+                % (default_product_plan.edition, plan_key.is_trial)
```
```suggestion
                f"Default for edition '{default_product_plan.edition}' "
                f"with is_trial='{plan_key.is_trial}' already exists."
```

biyeun commented on corehq/apps/accounting/tests/test_ensure_plans.py:
```diff
@@ -25,47 +26,47 @@ def tearDown(self):
         self._test_plan_versions_ensured(BOOTSTRAP_CONFIG_TESTING)
         self._test_plan_versions_ensured({
-            (SoftwarePlanEdition.COMMUNITY, False, False): {
+            (SoftwarePlanEdition.COMMUNITY, False, False, False): {
```
were any tests added for the annual plans?

biyeun replied:
nm I see in `BOOTSTRAP_CONFIG_TESTING` those annual plans are added

biyeun reviewed (approved):
thanks for the refactor as well!
left some comments on moving to f-strings to improve readability, but those aren't blocking

nospame committed:
Use f strings in ensure_plans for logging and name construction

Also remove duplicate logging for software product rate, and note whether default product plan is an annual plan.

nospame commented:
Want to confirm there are no particular concerns with this. Don't know that I've ever _not_ been able to check that box before.

> The migration here relies on code changes to ensure_plans to run correctly. Happy to split up this PR if the code changes are required to come first.

jingcheng16 commented:
@millerdev Hi Daniel, any thoughts on this?
https://github.com/dimagi/commcare-hq/pull/34584#issuecomment-2116317769
Migration related questions always stumped me.

millerdev reviewed (commented):
> The migration here relies on code changes to ensure_plans to run correctly.

Thanks for bringing this to my attention. I didn't notice it and did not plan to do a thorough review of this PR since Biyeun had already reviewed it.

With regard to:

> The migrations in this code can be safely applied first independently of the code

Another way to say this is, some currently-deployed code would be broken and cause errors if the migrations in this PR were to be applied before the PR is deployed. (This is what happens, for a usually short time, during every deploy.)

It's not clear to me (yet) if that is true for this PR, but it would be true, for example, in a PR that deleted a table and also removed code that referenced that same table. Those two things (_remove code that references to-be-deleted table_ and _migration that deletes a table_) need to be done in separate PRs (and deployed probably 6 weeks apart) so that there is no live code that references the table by the time the migration is applied. The 6-week constraint is so self-hosted environments have a chance to apply the migration before the breaking change is deployed.

On the other hand, it's fine for a single PR to include a migration that adds a new table and also include code that references that new table. This is safe since there is no deployed code that references the table.

Does this explanation help you form a better answer? Hopefully it was helpful and not a bunch of things you already knew.

nospame commented on corehq/apps/accounting/models.py:
```diff
@@ -857,7 +857,7 @@ class DefaultProductPlan(models.Model):
     class Meta(object):
         app_label = 'accounting'
-        unique_together = ('edition', 'is_trial', 'is_report_builder_enabled')
+        unique_together = ('edition', 'is_trial', 'is_report_builder_enabled', 'is_annual_plan')
```
> Another way to say this is, some currently-deployed code would be broken and cause errors if the migrations in this PR were to be applied before the PR is deployed.
...
> Does this explanation help you form a better answer?

Yes, I think so. I believe this is closer to "include a migration that adds a new table and also include code that references that new table", in that the migration first alters a table, and then relies on that altered table state to add new software plans.

> I'm not sure how much data is in that table

Quite little - essentially one object for every uniquely named Software Plan, in this case a quick query returns a total of 12 `DefaultProductPlan`s on production, 10 on staging, and 15 (after running this migration) in my local environment.

I'm not really familiar with indices. Would creating the new index `CONCURRENTLY` first mean we're generating a new set of data with the change applied, waiting for that to complete, and then pointing the db to the new data? It seems like this might not be necessary given the small volume of data, but is it better practice to just go this route anyway?

> Finally, are there any potential uniqueness violations in the existing data?

There are not. `DefaultProductPlan` is only created in `ensure_plans`, which itself is only run in migrations. The `is_annual_plan` property was only recently added, with a default of `False` that is now applied to all instances existing prior to this PR, and because of the current constraint could not violate the new one.

millerdev replied:
If there are so few (i.e., less than thousands) it will be no problem to do it this way, no need to do it `CONCURRENTLY`.

nospame committed:
Make migration reversible

nospame commented on corehq/apps/accounting/migrations/0094_add_annual_softwareplans.py:
```diff
@@ -9,6 +9,15 @@ def _bootstrap_new_annual_pricing(apps, schema_editor):
     ensure_plans(BOOTSTRAP_CONFIG, verbose=True, apps=apps)
 
 
+def _remove_annual_default_product_plans(apps, schema_editor):
```
I don't know if this is strictly necessary but was easy enough to add, so that _in case_ it needs to be reverted for any reason we won't be stuck doing so manually on every environment.

nospame commented on corehq/apps/accounting/bootstrap/utils.py:
```diff
@@ -117,38 +66,40 @@ def _ensure_product_rate(monthly_fee, edition, verbose, apps):
-        product_rate.name = product_name
+
+def _get_software_product(product_name, verbose, apps):
+    # SoftwareProduct no longer exists but is retained here to avoid breaking old migrations
```
I fully agree, but I'm not actually sure how to do that. `SoftwareProduct` is currently created in accounting migration `0001_squashed_0052_ensure_report_builder_plans`, modified in `0010_remove_softwareproduct_product_type` and deleted entirely in `0011_remove_softwareproduct`, all of which are ~7 years old. The similarly aged comment that stood here previously noted that this could be removed "after squashing migrations" -- I'm not sure if there was more squashing intended or if it was just not fully correct.
